types/ember/index.d.ts

```diff
@@ -342,10 +446,73 @@ export namespace Ember {
     and actions are targeted at the view object. There is no access to the surrounding context or
     outer controller; all contextual information is passed in.
     **/
-    class Component extends Object {
-        sendAction(action: string, context: any): void;
-        targetObject: Controller;
-        static positionalParams: string | string[];
+    class Component extends CoreView.extend(ViewMixin, ActionSupport, ClassNamesSupport) {
```

Comment:
Should there be support for dynamic/runtime keys on the component? There will obviously be a loss of type-safety since they will be typed as `{[key: string]: any}` or is the mandate that key's be static with a default value? I feel like the former is more in line with how Ember works `{{my-component dynamicPropKey=anyValue}}`,  but I think I might prefer the latter because of type-checking...

Reply:
Yeah, if you add the looser type you lose *all* help from the compiler for `get` and so on, alas.

For TS to be useful people will need to declare what the component expects as its API in the class definition for it, because while you *can* set arbitrary keys on a component invocation, they’re essentially meaningless: nothing will be done with them. And if they’re not part of the compnenet’s own interface, it certainly won’t be able to do anything with them.

For cases where people want to support adding arbitrary values at a given key, they can type that key as `variousData: any` in the component declaration and then it’ll be available within the component methods.